samcli/local/docker/container.py

```diff
@@ -296,7 +307,16 @@ def wait_for_result(self, full_path, event, stdout, stderr):
             self._logs_thread = threading.Thread(target=self.wait_for_logs, args=(stderr, stderr), daemon=True)
             self._logs_thread.start()
 
+        # wait_for_http_response will attempt to establish a connection to the socket
+        # but it'll fail if the socket is not listening yet, so we wait for the socket
+        self.wait_for_socket_connection()
+
+        # start the timer for function timeout right before executing the function, as waiting for the socket
+        # can take some time
+        timer = start_timer() if start_timer else None
         self.wait_for_http_response(full_path, event, stdout)
+        if timer:
+            timer.cancel()
```

Comment:
I had to update/move the timer logic (originally everything happened in runtime.py and we started the timer when we called `configure_interrupt`) in order to start the timer after waiting for the socket, because the default function timeout is 3 seconds and waiting for the socket can take this amount of time (so it would timeout before even executing/invoking the function).